### PR TITLE
fix(angular): remove numbers from the project name

### PR DESCRIPTION
**Pull request by matheo**
## Current Behavior

Any number in the directory path can break the Angular CLI application schematic:
```bash
SchematicsException [Error]: Project name "src-9-public-website" is not valid.
New project names must start with a letter, and must contain only alphanumeric characters or dashes.
When adding a dash the segment after the dash must also start with a letter.
src-9-public-website
                  ^
    at Object.validateProjectName (node_modules/@schematics/angular/utility/validation.js:45:15)
```

## Expected Behavior

The schematic should be able to run without exceptions even if numbers are present in the `options.directory`.

### CODE REVIEW

File: packages/angular/src/generators/application/lib/normalize-options.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@ export function normalizeOptions(
     : names(options.name).fileName;
 
   let e2eProjectName = `${names(options.name).fileName}-e2e`;
-  const appProjectName = appDirectory.replace(new RegExp('/', 'g'), '-');
+  const appProjectName = appDirectory.replace(/[\/\d]/g, '-').replace(/[-]+/g, '-');
   if (options.e2eTestRunner !== 'cypress') {
     e2eProjectName = `${appProjectName}-e2e`;
   }
```
Comment (matheo) [inline packages/angular/src/generators/application/lib/normalize-options.ts:23]: `appsDir` cannot be empty?
I want to right-click a folder and run the application/library schematic with the directory relative to the root
so I've set empty `appsDir` and `libsDir` in the `nx.json` `workspaceLayout` but it adds a leading `/` because the following lines doesn't use `path.join`:

File: packages/angular/src/generators/application/lib/normalize-options.ts [hunk 2]
```diff
@@ -17,13 +18,15 @@ export function normalizeOptions(
...
+    .replace(/[\/\d]/g, '-')
+    .replace(/[-]+/g, '-');
   if (options.e2eTestRunner !== 'cypress') {
     e2eProjectName = `${appProjectName}-e2e`;
   }
 
-  const appProjectRoot = `${appsDir}/${appDirectory}`;
-  const e2eProjectRoot = `${appsDir}/${appDirectory}-e2e`;
+  const appProjectRoot = join(appsDir, appDirectory);
+  const e2eProjectRoot = join(appsDir, `${appDirectory}-e2e`);
```
Comment (matheo) [inline packages/angular/src/generators/application/lib/normalize-options.ts:29]: `appsDir` cannot be empty?
I want to right-click a folder and run the application/library schematic with the directory relative to the root
so I've set empty `appsDir` and `libsDir` in the `nx.json` workspaceLayout but it adds a leading `/` because these lines didn't use `path.join`

**Comment by matheo** [comment]
@FrozenPandaz can I get any eyes here please?

File: packages/angular/src/generators/application/lib/normalize-options.ts [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import type { Tree } from '@nrwl/devkit';
+import { join } from 'path';
```
Comment (Coly010) [inline packages/angular/src/generators/application/lib/normalize-options.ts:2]: Let's use `joinPathFragments` from `@nrwl/devkit`

**Review by Coly010 — CHANGES_REQUESTED** [review]
Thanks for the PR!

Can you add a unit test to validate that `src-9-directory` will be caught as invalid?

**Comment by matheo** [comment]
Hi @Coly010! I'm back to this one
I've found that there are a couple paths where empty `appsDir` and `libsDir` outputs a double slash.
I will find them and include them here, but please, can I ask your help to test a library and application inside a path like:
```
npx nx generate @nrwl/angular:library --name=button --directory=src/3-design-system
npx nx generate @nrwl/angular:application --name=demo --directory=src/9-websites
```
I'm really unfamiliar with the setup :(

File: packages/angular/src/generators/library/lib/update-tsconfig.ts [hunk 4]
```diff
@@ -14,9 +19,11 @@ function updateRootConfig(host: Tree, options: NormalizedSchema) {
...
 
     c.paths[options.importPath] = [
-      `${getWorkspaceLayout(host).libsDir}/${
-        options.projectDirectory
-      }/src/index.ts`,
+      joinPathFragments(
+        getWorkspaceLayout(host).libsDir,
+        options.projectDirectory,
+        '/src/index.ts'
+      ),
```
Comment (matheo) [inline packages/angular/src/generators/library/lib/update-tsconfig.ts:26]: this is a big deal because my tsconfig.base was configured with absolute routes due to the empty `libsDir` and the paths are not detected

**Comment by matheo** [comment]
@Coly010 done!
I've added the validations covered by this PR!
I'm so excited with Nx, looking forward to play with the source and contribute more :wink:

**Comment by matheo** [comment]
Is there a way to fix the e2e tests?
it seems like a broken dependency, is it the CI cache?

**Comment by Coly010** [comment]
Re-running the e2es. If they fail again, they should have a better output

File: packages/devkit/src/tests/create-tree-with-empty-workspace.ts [hunk 5]
```diff
@@ -4,7 +4,7 @@ import type { Tree } from '@nrwl/tao/src/shared/tree';
 /**
  * Creates a host for testing.
  */
-export function createTreeWithEmptyWorkspace(version = 1): Tree {
+export function createTreeWithEmptyWorkspace(version = 1, nxConfig = {}): Tree {
```
Comment (leosvelperez) [inline packages/devkit/src/tests/create-tree-with-empty-workspace.ts:7]: Please revert this change. This helper function is meant to create a standard empty workspace.

Instead, after creating the empty tree in your tests, do:

```typescript
updateJson(tree, 'nx.json', json => {
  // modify the json
  return json;
})
```

**Review by leosvelperez — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Overall the changes look good to me. I only have one last change request.

The e2e issue in the pipeline should be solved with https://github.com/nrwl/nx/pull/7887. After it is merged, we need to rebase this one to ensure the e2e tests are good.

File: packages/devkit/src/tests/create-tree-with-empty-workspace.ts (hunk 5 above)
Comment (matheo) [reply packages/devkit/src/tests/create-tree-with-empty-workspace.ts]: done, seems good!

**Review by leosvelperez — APPROVED** [review]
LGTM! 🎉

**Comment by matheo** [comment]
@Coly010 can we merge this please?
or cannot be included in the `beta` despite it's not a feature but a patch?

**Comment by Coly010** [comment]
It will be released with next version of Nx, which should be 13.3 and should be this week
